### PR TITLE
qos support

### PR DESCRIPTION
**Pull request by vtolstov**

### CODE REVIEW

File: qos.go [hunk 1]
```diff
@@ -114,25 +114,47 @@ func (odbi *ovnDBImp) addQoSImp(ls string, direction string, priority *int, matc
 }
 
 func (odbi *ovnDBImp) delQoSImp(ls string, direction string, priority *int, match string) (*OvnCommand, error) {
```
Comment (softwarejl) [inline qos.go]: why using point?

**Comment by vtolstov** [comment]
@softwarejl pointer used to distinguish 0 vs unset param. Without it we always need to create rule with priority 0 if user not specify, i don't know what right solution....

**Comment by vtolstov** [comment]
> @softwarejl pointer used to distinguish 0 vs unset param. Without it we always need to create rule with priority 0 if user not specify, i don't know what right solution....

my mistake, this is mandatory param, so pointer not needed. i'm fix it

File: qos.go [hunk 2]
```diff
@@ -16,12 +16,188 @@
...
+		Table:     tableLogicalSwitch,
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{condition},
+	}
+	operations := []libovsdb.Operation{insertOp, mutateOp}
+
+	return &OvnCommand{operations, odbi, make([][]map[string]interface{}, len(operations))}, nil
+}
+
+func (odbi *ovnDBImp) delQoSImp(ls string, direction string, priority *int, match string) (*OvnCommand, error) {
```
Comment (softwarejl) [inline qos.go]: why priority is point?  no default value?

Comment (vtolstov) [reply qos.go]: without pointer we can't allow wildcard removing of qos rules like ovn-nbctl qos-del allows, so we need pointer in this place.

File: qos.go [hunk 3]
```diff
@@ -16,12 +16,188 @@
...
+
+	odbi.cachemutex.RLock()
+	defer odbi.cachemutex.RUnlock()
+
+	cacheLogicalSwitch, ok := odbi.cache[tableLogicalSwitch]
+	if !ok {
+		return nil, ErrorNotFound
+	}
+
+	for _, drows := range cacheLogicalSwitch {
```
Comment (softwarejl) [inline qos.go]: is it possible to have common function for that, so many duplicated code now. 

interface := drows.Fields["xxxx"]
interface.(type)
switch.....

not sure it works.

Comment (vtolstov) [reply qos.go]: i'm try to think about this, but now i don't know how to do that.

File: qos.go (hunk 2 above)
Comment (softwarejl) [reply qos.go]: OK.

File: acl.go [hunk 4]
```diff
@@ -42,30 +48,34 @@ func (odbi *ovnDBImp) getACLUUIDByRow(lsw, table string, row OVNRow) (string, er
...
 							if va, ok := a.(libovsdb.UUID); ok {
+								cacheACL, ok := odbi.cache[tableACL][va.GoUUID]
+								if !ok {
+									return "", ErrorNotFound
+								}
 								for field, value := range row {
 									switch field {
 									case "action":
-										if odbi.cache[tableACL][va.GoUUID].Fields["action"].(string) != value {
+										if cacheACL.Fields["action"].(string) != value {
```
Comment (softwarejl) [inline acl.go]: in future, maybe have head file to define the table structure, and here can be simplified to (TABLE field name), (TABLE field type).

thoughts?

Comment (vtolstov) [reply acl.go]: i'm prefer to try to utilize hashicorp/mapstructure as row have map[string]interface{} format we can automatic unmrashal this row to struct, for types like OvsSet and other specific we can use custom decoder...
So huge part of code can be replaced.

**Comment by vtolstov** [comment]
@softwarejl i have trouble for qos deletion, as i see ovn-nbctl send this request:
```
2019-03-22T17:05:34Z|00024|jsonrpc|DBG|unix:/var/run/openvswitch/ovnnb_db.sock: send request, method="transact", params=["OVN_Northbound",{"rows":[{"qos_rules":["set",[["uuid","199ba4ba-8cc6-44ef-abab-42b94d078650"],["uuid","25695f8d-d1e3-4381-9973-adc841b3a00c"],["uuid","f20a496c-440f-4fa7-abaf-28b7743699dc"]]]}],"until":"==","where":[["_uuid","==",["uuid","83e587e9-00da-42e3-b0a8-631174ae71c8"]]],"columns":["qos_rules"],"timeout":0,"op":"wait","table":"Logical_Switch"},{"where":[["_uuid","==",["uuid","83e587e9-00da-42e3-b0a8-631174ae71c8"]]],"row":{"qos_rules":["set",[]]},"op":"update","table":"Logical_Switch"},{"comment":"ovs-nbctl: ovn-nbctl -v qos-del TEST_LSW to-lport","op":"comment"}], id=4
```

I'm delete 3 qos rules from logical switch.

File: acl.go (hunk 4 above)
Comment (softwarejl) [reply acl.go]: same idea with me,  could we have issue ticket to track it and discuss.

Comment (vtolstov) [reply acl.go]: https://github.com/eBay/go-ovn/issues/29 needs more generic name..

**Comment by vtolstov** [comment]
@softwarejl @hzhou8 as i see qos test run without errors now, and api mimics ovn-nbctl.

**Comment by vtolstov** [comment]
i'm rebase and aff comments

**Pull request merged**